api/admin/controller.py

```diff
@@ -805,6 +807,81 @@ def genres(self):
             })
         return data
 
+class CustomListsController(CirculationManagerController):
+    def custom_lists(self):
+        library = flask.request.library
+
+        data_source = DataSource.lookup(self._db, DataSource.LIBRARY_STAFF)
+
+        if flask.request.method == "GET":
+            custom_lists = []
+            for list in library.custom_lists:
+                entries = []
+                for entry in list.entries:
+                    entries.append(dict(id=entry.edition.primary_identifier.urn,
+                                        title=entry.edition.title,
+                                        authors=[author.display_name for author in entry.edition.author_contributors],
+                                        ))
+                custom_lists.append(dict(id=list.id, name=list.name, entries=entries))
+            return dict(custom_lists=custom_lists)
+
+        if flask.request.method == "POST":
+            id = flask.request.form.get("id")
+            name = flask.request.form.get("name")
+            entries = flask.request.form.get("entries")
+
+            if id:
+                is_new = False
+                list = get_one(self._db, CustomList, id=int(id), data_source=data_source)
+                if not list:
+                    return MISSING_CUSTOM_LIST
+                if list.library != library:
+                    return CANNOT_CHANGE_LIBRARY_FOR_CUSTOM_LIST
+            else:
+                list, is_new = create(self._db, CustomList, name=name, data_source=data_source)
+                list.created = datetime.now()
+                list.library = library
+
+            list.updated = datetime.now()
+            list.name = name
+
+            if entries:
+                entries = json.loads(entries)
+            else:
+                entries = []
+
+            old_entries = list.entries
+            for entry in entries:
+                identifier, ignore = Identifier.parse_urn(self._db, entry.get("id"))
+                pools = self._db.query(
+                    LicensePool
+                ).join(
+                    LicensePool.collection
+                ).join(
+                    LicensePool.identifier
+                ).join(
+                    Collection.libraries
+                ).filter(
+                    Identifier.id==identifier.id
+                ).filter(
+                    Library.id==library.id
+                ).all()
+                if not pools:
+                    return CANNOT_ADD_CUSTOM_LIST_ENTRY_WITHOUT_LICENSEPOOL
+
+                edition = pools[0].presentation_edition
```

Comment:
I'm not sure if using the first pool is the best way to do this...

Reply:
This is where switching to Works will pay off.